angel_system/impls/detect_activities/swinb/swinb_detect_activities.py

```diff
@@ -0,0 +1,209 @@
+    """
+
+    def __init__(
+        self,
+        checkpoint_path: str,
+        num_classes: int,
+        labels_file: str,
+        num_frames: int = 32,
+        sampling_rate: int = 2,
+        torch_device: str = "cpu",
+        det_threshold: float = 0.75,
+    ):
+        self._checkpoint_path = checkpoint_path
+        self._num_classes = num_classes
+        self._torch_device = torch_device
+        self._det_threshold = det_threshold
+        self._num_frames = num_frames
+        self._sampling_rate = sampling_rate
+        self._labels_file = labels_file
+
+        # Set to None for lazy loading later.
+        self._model: torch.nn.Module = None  # type: ignore
+        self._model_device: torch.device = None  # type: ignore
+
+        # Default configs from learn SwinVideo config
+        self._mean = [0.45, 0.45, 0.45]
+        self._std = [0.225, 0.225, 0.225]
+        self._crop_size = 224
+        self._frames_per_second = 30
+
+        # Transfrom from learn/TimeSformer/video_classification.py
+        self.transform = torchvision.transforms.Compose(
+            [
+                torchvision.transforms.ToTensor(),
+                torchvision.transforms.Normalize(
+                    self._mean, self._std
+                )
+            ]
+        )
+
+        # Set up the labels from the given labels file
+        self._labels = []
+        with open(self._labels_file, "r") as f:
+            for line in f:
+                self._labels.append(line.rstrip())
+
+    def get_model(self) -> torch.nn.Module:
+        """
+        Lazy load the torch model in an idempotent manner.
+        :raises RuntimeError: Use of CUDA was requested but is not available.
+        """
+        model = self._model
+        if model is None:
+            # Load the model with the checkpoint
+            model = swin_b(self._checkpoint_path, self._num_classes)
+            model = model.eval()
+
+            # Transfer the model to the requested device
+            if self._torch_device != 'cpu':
+                if torch.cuda.is_available():
+                    model_device = torch.device(device=self._torch_device)
+                    model = model.to(device=model_device)
+                else:
+                    raise RuntimeError(
+                        "Use of CUDA requested but not available."
+                    )
+            else:
+                model_device = torch.device(self._torch_device)
+
+            self._model = model
+            self._model_device = model_device
+
+        return model
+
+    def detect_activities(
+        self,
+        frame_iter: Iterable[np.ndarray]
+    ) -> Iterable[str]:
+        """
+        Formats the given iterable of frames into the required input format
+        for the swin model and then inputs them to the model for inferencing.
+        """
+        # Check that we got the right number of frames
+        assert len(list(frame_iter)) == (self._sampling_rate * self._num_frames)
```

Comment:
This cast is incorrectly located: if `frame_iter` was truly an iterable, then this would "consume" it and later iterations would yield nothing. What I meant was to capture the list-ification, and then use that throughout this method body:
```suggestion
        frame_iter = list(frame_iter)
        assert len(frame_iter) == (self._sampling_rate * self._num_frames)
```